lib/grpc/adapter/gun.ex

```diff
@@ -12,24 +12,26 @@ defmodule GRPC.Adapter.Gun do
   def connect(channel, opts), do: connect_insecurely(channel, opts)
 
   defp connect_securely(%{cred: %{ssl: ssl}} = channel, opts) do
-    transport_opts = Map.get(opts, :transport_opts, @default_transport_opts ++ ssl)
-    open_opts = %{transport: :ssl, protocols: [:http2], transport_opts: transport_opts}
+    transport_opts = Map.get(opts, :transport_opts, @default_transport_opts)
+    open_opts = %{transport: :ssl, protocols: [:http2], tcp_opts: transport_opts, tls_opts: ssl}
     open_opts = Map.merge(opts, open_opts)
 
     do_connect(channel, open_opts)
   end
 
   defp connect_insecurely(channel, opts) do
     transport_opts = Map.get(opts, :transport_opts, @default_transport_opts)
-    open_opts = %{transport: :tcp, protocols: [:http2], transport_opts: transport_opts}
+    open_opts = %{transport: :tcp, protocols: [:http2], tcp_opts: transport_opts}
     open_opts = Map.merge(opts, open_opts)
 
     do_connect(channel, open_opts)
   end
 
   defp do_connect(%{host: host, port: port} = channel, open_opts) do
     open_opts =
-      Map.merge(%{retry: :infinity, retry_timeout: &GRPC.Stub.retry_timeout/1}, open_opts)
+      Map.merge(%{retry: 1, retry_fun: fn(retries, _) ->
+        %{retries: retries + 1, timeout: GRPC.Stub.retry_timeout(retries)}
+      end}, open_opts)
```

Comment:
Hi @essen, @tony612 
I am trying to speed up the progress of this PR.
I tested a little bit the code and trace the gun implementation. When you set `retry: 1`, it does not retry actually. It is minus 1 before pass to `not_connected` function. So it causes some test cases failed. https://github.com/ninenines/gun/blob/master/src/gun.erl#L1439

To make the behavior keep the same as before, ie. "retry forever and increase timeout after each retry", the retry should be set to 2. But I think this code didn't show the intention well.

Reply:
Sounds like a bug in Gun, if retry is set to 1 it should retry at least once. I will look into it.

Reply:
Thanks for the debugging by the way!

Reply:
Please try the most recent Gun I should have fixed the issue.

Reply:
Cool, it's working at my local, thanks! 
In case if you are not familiar with elixir tools, you could run `mix deps.update gun` to update `mix.lock`, then push the changes. Hopefully, it should pass the CI test. 😄

Reply:
Hi @essen, do you have any release schedule about the cowboy/gun? thanks

Reply:
Cowboy 2.7 is next week. Gun 2.0 probably at the end of the year, but I'd recommend using a pre-release before that if that's OK for the grpc project. I can issue a pre-release with fixed dependencies if necessary. The blocker for Gun 2.0 is HTTP/2 CONNECT/Websocket which might or might not require some public interface changes.

Reply:
cool, thanks, I think the cowboy is more commonly used, so it helps a lot.

btw, I try to update the cowboy to the latest commit: cc54c207e35f3ab7a2dfc105eef39fe7d3bf1633, but the test failed.  I didn't figure out the root cause yet. The server process seems to be blocked in https://github.com/ninenines/cowboy/blob/cc54c207e35f3ab7a2dfc105eef39fe7d3bf1633/src/cowboy_req.erl#L522 until timeout.

Reply:
Don't forget to also update Cowlib.

Reply:
If you could try tracing to see what's happening that would help, I can't look right now. Which test is it by the way? It could just be that the HTTP/2 CVE fixes get in the way of the tests, as was the case in some of Cowboy's own test suite.